staker/bold_state_provider.go

```diff
@@ -0,0 +1,498 @@
+// Copyright 2023, Offchain Labs, Inc.
+// For license information, see https://github.com/offchainlabs/bold/blob/main/LICENSE
+package staker
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/metrics"
+
+	protocol "github.com/OffchainLabs/bold/chain-abstraction"
+	"github.com/OffchainLabs/bold/containers/option"
+	l2stateprovider "github.com/OffchainLabs/bold/layer2-state-provider"
+	"github.com/OffchainLabs/bold/state-commitments/history"
+
+	"github.com/offchainlabs/nitro/arbutil"
+	challengecache "github.com/offchainlabs/nitro/staker/challenge-cache"
+	"github.com/offchainlabs/nitro/validator"
+)
+
+var (
+	_ l2stateprovider.ProofCollector          = (*BOLDStateProvider)(nil)
+	_ l2stateprovider.L2MessageStateCollector = (*BOLDStateProvider)(nil)
+	_ l2stateprovider.MachineHashCollector    = (*BOLDStateProvider)(nil)
+	_ l2stateprovider.ExecutionProvider       = (*BOLDStateProvider)(nil)
+)
+
+var executionNodeOfflineGauge = metrics.NewRegisteredGauge("arb/state_provider/execution_node_offline", nil)
+
+var (
+	ErrChainCatchingUp = errors.New("chain catching up")
+)
+
+type BOLDStateProvider struct {
+	validator            *BlockValidator
+	statelessValidator   *StatelessBlockValidator
+	historyCache         challengecache.HistoryCommitmentCacher
+	challengeLeafHeights []l2stateprovider.Height
+	validatorName        string
+	checkBatchFinality   bool
+	sync.RWMutex
+}
+
+type BOLDStateProviderOpt = func(b *BOLDStateProvider)
+
+func WithoutFinalizedBatchChecks() BOLDStateProviderOpt {
+	return func(b *BOLDStateProvider) {
+		b.checkBatchFinality = false
+	}
+}
+
+func NewBOLDStateProvider(
+	blockValidator *BlockValidator,
+	statelessValidator *StatelessBlockValidator,
+	cacheBaseDir string,
+	challengeLeafHeights []l2stateprovider.Height,
+	validatorName string,
+	opts ...BOLDStateProviderOpt,
+) (*BOLDStateProvider, error) {
+	historyCache, err := challengecache.New(cacheBaseDir)
+	if err != nil {
+		return nil, err
+	}
+	sp := &BOLDStateProvider{
+		validator:            blockValidator,
+		statelessValidator:   statelessValidator,
+		historyCache:         historyCache,
+		challengeLeafHeights: challengeLeafHeights,
+		validatorName:        validatorName,
+		checkBatchFinality:   true,
+	}
+	for _, o := range opts {
+		o(sp)
+	}
+	return sp, nil
+}
+
+// Produces the L2 execution state to assert to after the previous assertion state.
+// Returns either the state at the batch count maxInboxCount or the state maxNumberOfBlocks after previousBlockHash,
+// whichever is an earlier state. If previousBlockHash is zero, this function simply returns the state at maxInboxCount.
+// TODO: Check the block validator has validated the execution state we are proposing.
```

Comment:
I have reworded the comment to make it a bit more clearer. 
I think we had it as separate functions earlier but then we had too many api b/w Bold and nitro.
Though I guess we could revisit this decision again considering the current state.
@rauljordan cc

Reply:
@rauljordan do I underand correctly that previousBlockGlobalState = nil only when there is no previous correct assertion? (as seems to be suggestion because fromBatch = 0)?
If so, shouldn't we still be limited by maxNumberOfBlocks?
Or are there two ways to post the same assertion (either from prev. confirmed or from 0)?
This should be clarified, at least in comments.

Reply:
Hey guys, I think this needs to be reworked as it is doing too much. One of the complex parts of this is that we need to:
1. Produce the execution state that follows a previous state, as the prev state determines how many inbox messages the next state should consume
2. Be careful around the max number of blocks limit, and if we reach it, we need to provide a correct assertion for that
This function was an attempt to push less responsibility onto the caller, but we might want to retrace that for the sake of safety and clarity

Reply:
@eljobe WDYT?